service/storage_proxy.cc

```diff
@@ -558,16 +626,21 @@ class storage_proxy::remote {
         return _sp.container().invoke_on(shard, _sp._write_ack_smp_service_group,
                 [from, response_id, num_failed, backlog = std::move(backlog), exception = std::move(exception)] (storage_proxy& sp) mutable {
             error err = error::FAILURE;
+            std::optional<sstring> msg;
             if (exception) {
-                err = std::visit([] <typename Ex> (Ex&) {
+                err = std::visit([&] <typename Ex> (Ex& e) {
                     if constexpr (std::is_same_v<Ex, replica::rate_limit_exception>) {
+                        msg = e.what();
                         return error::RATE_LIMIT;
                     } else if constexpr (std::is_same_v<Ex, replica::unknown_exception> || std::is_same_v<Ex, replica::no_exception>) {
                         return error::FAILURE;
+                    } else if constexpr(std::is_same_v<Ex, replica::stale_topology_exception>) {
+                        msg = e.what();
+                        return error::FAILURE;
```

Comment:
Ah, I see you do use exceptions-as-values. As long as it's done wherever possible, I won't ask to extend it to where it's not used yet.

Reply:
But, perhaps we should allocate a new error code.